torch/nn/modules/container.py

```diff
@@ -457,85 +451,81 @@ def __getitem__(self: T, idx: slice) -> T:
         if isinstance(idx, slice):
-            return self.__class__(list(self._parameters.values())[idx])
+            start, stop, step = idx.indices(len(self))
+            out = self.__class__()
+            for i in range(start, stop, step):
+                out.append(self[i])
+            return out
         else:
             idx = self._get_abs_string_index(idx)
-            return self._parameters[str(idx)]
+            return getattr(self, str(idx))
 
-    def __setitem__(self, idx: int, param: 'Parameter') -> None:
+    def __setitem__(self, idx: int, param: Any) -> None:
         idx = self._get_abs_string_index(idx)
-        return self.register_parameter(str(idx), param)
-
-    def __setattr__(self, key: Any, value: Any) -> None:
-        if getattr(self, "_initialized", False):
-            if not hasattr(self, key) and not isinstance(value, torch.nn.Parameter):
-                warnings.warn("Setting attributes on ParameterList is not supported.")
-        super(ParameterList, self).__setattr__(key, value)
+        return setattr(self, str(idx), param)
 
     def __len__(self) -> int:
-        return len(self._parameters)
+        return self._size
 
-    def __iter__(self) -> Iterator['Parameter']:
-        return iter(self._parameters.values())
+    def __iter__(self) -> Iterator[Any]:
+        return iter(self[i] for i in range(len(self)))
 
-    def __iadd__(self, parameters: Iterable['Parameter']) -> 'ParameterList':
+    def __iadd__(self, parameters: Iterable[Any]) -> 'ParameterList':
         return self.extend(parameters)
 
     def __dir__(self):
-        keys = super(ParameterList, self).__dir__()
-        keys = [key for key in keys if not key.isdigit()]
-        return keys
+        return list(range(self._size))
 
-    def append(self, parameter: 'Parameter') -> 'ParameterList':
-        """Appends a given parameter at the end of the list.
+    def append(self, value: Any) -> 'ParameterList':
+        """Appends a given value at the end of the list.
 
         Args:
-            parameter (nn.Parameter): parameter to append
+            value (Any): value to append
         """
-        self.register_parameter(str(len(self)), parameter)
+        setattr(self, str(len(self)), value)
+        self._size += 1
         return self
 
-    def extend(self, parameters: Iterable['Parameter']) -> 'ParameterList':
-        """Appends parameters from a Python iterable to the end of the list.
+    def extend(self, values: Iterable[Any]) -> 'ParameterList':
+        """Appends values from a Python iterable to the end of the list.
 
         Args:
-            parameters (iterable): iterable of parameters to append
+            values (iterable): iterable of values to append
         """
-        if not isinstance(parameters, container_abcs.Iterable):
+        # Tensor is an iterable but we never want to unpack it here
+        if not isinstance(values, container_abcs.Iterable) or isinstance(values, torch.Tensor):
```

Comment:
do we want `torch.overrides.is_tensor_like()` here instead?

Reply:
That is a broader `torch.nn` question I think.  Leaving this as-is for now?